File: Top-Interview-Questions-Leetcode/Easy-Collections/strings/reverse-integer.cpp
#include <iostream>
#include <vector>
#include <string>

// Use this if you want to use the optimized version
// #define OPTIMIZED

class Solution {
private:
    
    int compare_string(const std::string& num1, const std::string& num2) {        
        if (num1.size() != num2.size()) {
            return num1.size() > num2.size() ? 1 : -1;
        }
        
        if (num1 == num2) return 0;
        
        return num1 > num2 ? 1 : -1;
    }
    
    void reverse(std::string& str) {
        int l = 0, h = str.size() - 1;
        
        while (l < h) {
            std::swap(str[l++], str[h--]);
        }
    }
    
public:
    int reverse(int x) {
        
        #ifdef OPTIMIZED
        return reverse_optimized(x);
        #endif
        
        std::string sx = std::to_string(x);
        
        if (x < 0) sx = sx.substr(1);
        
        reverse(sx);
        
        std::string inf = std::to_string((1 << 31)), sup = std::to_string(-((1 << 31) + 1));
        
        inf = inf.substr(1);
        
        if (compare_string(sx, inf) > 0 or compare_string(sx, sup) > 0) return 0;
        
        if (x > 0) {
            return stoi(sx);
        }
        else return - stoi(sx);
    }
    
    int reverse_optimized(int x) {
        int reverse_x = 0;
        
        while (x != 0) {
            int digit = x % 10;
            
            if (INT_MAX / 10 < reverse_x or (INT_MAX / 10 == reverse_x and digit > 7)) return 0;
            if (INT_MIN / 10 > reverse_x or (INT_MIN / 10 == reverse_x and digit < -8)) return 0;
            
            reverse_x = 10 * reverse_x + digit;
            x /= 10;
        }
        
        return reverse_x;
    }
};

int main() {
    Solution solution;
    
    std::cout << solution.reverse(123) << '\n'; 
    std::cout << solution.reverse(-123) << '\n'; 
    std::cout << solution.reverse(120) << '\n'; 
    std::cout << solution.reverse(0) << '\n'; 
    std::cout << solution.reverse(1534236469) << '\n';
    
    return 0;
}

File: Top-Interview-Questions-Leetcode/Easy-Collections/strings/reverse-integer.py
class Solution:
    def reverse(self, x: int) -> int:
        sign = -1 if x < 0 else 1
        
        x = abs(x)
        
        reversed_x = 0
        
        INT_MAX = (1 << 31) - 1
        INT_MIN = - (1 << 31)

        while x > 0:
            digit = x % 10
            
            if reversed_x > INT_MAX // 10 or (reversed_x == INT_MAX // 10 and digit > 7):
                return 0
            
            reversed_x = reversed_x * 10 + digit
            x //= 10

        result = sign * reversed_x
        
        if result > INT_MAX or result < INT_MIN:
            return 0
        
        return result

# Example usage
def main():
    solution = Solution()
    test_cases = [123, -123, 120, 0, 1534236469]
    
    for x in test_cases:
        result = solution.reverse(x)
        print(f"Input: {x}, Reversed: {result}")

if __name__ == "__main__":
    main()